File: src/main/java/agentdecompile/util/MemoryUtil.java
/* ###
 * IP: AgentDecompile
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package agentdecompile.util;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Program;
import ghidra.program.model.mem.Memory;
import ghidra.program.model.mem.MemoryAccessException;
import ghidra.program.model.mem.MemoryBlock;

/**
 * Utility functions for working with Ghidra memory.
 * <p>
 * Ghidra Memory API references:
 * <ul>
 *   <li>{@link ghidra.program.model.mem.Memory} - <a href="https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/Memory.html">Memory API</a></li>
 *   <li>{@link ghidra.program.model.mem.MemoryBlock} - <a href="https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/MemoryBlock.html">MemoryBlock API</a></li>
 *   <li>{@link ghidra.program.model.mem.MemoryAccessException} - <a href="https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/MemoryAccessException.html">MemoryAccessException API</a></li>
 * </ul>
 * See <a href="https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/package-summary.html">ghidra.program.model.mem package</a>.
 * </p>
 */
public class MemoryUtil {

    /**
     * Format a byte array as a hex string
     * @param bytes The byte array
     * @return A hex string representation
     */
    public static String formatHexString(byte[] bytes) {
        if (bytes == null || bytes.length == 0) {
            return "";
        }

        StringBuilder hexBuilder = new StringBuilder();
        for (byte b : bytes) {
            hexBuilder.append(String.format("%02X ", b & 0xFF));
        }
        return hexBuilder.toString().trim();
    }

    /**
     * Format a byte array as an ASCII string (non-printable chars shown as '.')
     * @param bytes The byte array
     * @return An ASCII string representation
     */
    public static String formatAsciiString(byte[] bytes) {
        if (bytes == null || bytes.length == 0) {
            return "";
        }

        StringBuilder asciiBuilder = new StringBuilder();
        for (byte b : bytes) {
            char c = (char) (b & 0xFF);
            if (c >= 32 && c <= 126) {
                asciiBuilder.append(c);
            } else {
                asciiBuilder.append('.');
            }
        }
        return asciiBuilder.toString();
    }

    /**
     * Convert a byte array to a list of integer values (0-255)
     * @param bytes The byte array
     * @return List of integer values
     */
    public static List<Integer> byteArrayToIntList(byte[] bytes) {
        if (bytes == null || bytes.length == 0) {
            return List.of();
        }

        List<Integer> result = new ArrayList<>(bytes.length);
        for (byte b : bytes) {
            result.add(b & 0xFF);
        }
        return result;
    }

    /**
     * Read memory bytes safely
     * @param program The Ghidra program
     * @param address Starting address
     * @param length Number of bytes to read
     * @return Byte array or null if an error occurred
     */
    public static byte[] readMemoryBytes(Program program, Address address, int length) {
        // Ghidra API: Program.getMemory() - https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Program.html#getMemory()
        Memory memory = program.getMemory();
        byte[] bytes = new byte[length];

        try {
            // Ghidra API: Memory.getBytes(Address, byte[]) - https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/Memory.html#getBytes(ghidra.program.model.address.Address,byte[])
            int read = memory.getBytes(address, bytes);
            if (read != length) {
                byte[] actualBytes = new byte[read];
                System.arraycopy(bytes, 0, actualBytes, 0, read);
                return actualBytes;
            }
            return bytes;
        } catch (MemoryAccessException e) {
            return null;
        }
    }

    /**
     * Find a memory block by name
     * @param program The Ghidra program
     * @param blockName Name of the block to find
     * @return The memory block or null if not found
     */
    public static MemoryBlock findBlockByName(Program program, String blockName) {
        // Ghidra API: Program.getMemory() - https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Program.html#getMemory()
        Memory memory = program.getMemory();
        // Ghidra API: Memory.getBlocks() - https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/Memory.html#getBlocks()
        for (MemoryBlock block : memory.getBlocks()) {
            // Ghidra API: MemoryBlock.getName() - https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/MemoryBlock.html#getName()
            if (block.getName().equals(blockName)) {
                return block;
            }
        }
        return null;
    }

    /**
     * Find the memory block containing the given address
     * @param program The Ghidra program
     * @param address The address to look up
     * @return The memory block or null if not found
     */
    public static MemoryBlock getBlockContaining(Program program, Address address) {
        // Ghidra API: Program.getMemory() - https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Program.html#getMemory()
        Memory memory = program.getMemory();
        // Ghidra API: Memory.getBlock(Address) - https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/Memory.html#getBlock(ghidra.program.model.address.Address)
        return memory.getBlock(address);
    }

    /**
     * Process memory bytes in chunks to avoid large memory allocations
     * @param program The Ghidra program
     * @param startAddress Starting address
     * @param length Total number of bytes to process
     * @param chunkSize Maximum chunk size
     * @param processor Consumer function that processes each chunk
     */
    public static void processMemoryInChunks(
            Program program,
            Address startAddress,
            long length,
            int chunkSize,
            Consumer<byte[]> processor) {

        // Ghidra API: Program.getMemory() - https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Program.html#getMemory()
        Memory memory = program.getMemory();
        Address currentAddress = startAddress;
        long remaining = length;

        while (remaining > 0) {
            int currentChunkSize = (int) Math.min(remaining, chunkSize);
            byte[] buffer = new byte[currentChunkSize];

            try {
                // Ghidra API: Memory.getBytes(Address, byte[]) - https://ghidra.re/ghidra_docs/api/ghidra/program/model/mem/Memory.html#getBytes(ghidra.program.model.address.Address,byte[])
                int read = memory.getBytes(currentAddress, buffer);
                if (read > 0) {
                    processor.accept(buffer);
                    // Ghidra API: Address.add(long) - https://ghidra.re/ghidra_docs/api/ghidra/program/model/address/Address.html#add(long)
                    currentAddress = currentAddress.add(read);
                    remaining -= read;
                } else {
                    break; // Could not read any bytes
                }
            } catch (MemoryAccessException e) {
                break; // Memory access error
            }
        }
    }
}

File: src/agentdecompile_cli/mcp_utils/memory_util.py
"""Memory utility helpers for safe reads and block inspection.

Used by inspect-memory/read-bytes and other providers to: read bytes at an
address with bounds checking (contains), convert Java byte arrays to Python
bytes, and format hex dumps. MemoryUtil.read_memory_bytes is the safe entry
point when the provider cannot assume getBytes/getByte always succeeds.
"""

from __future__ import annotations

import logging

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

    from ghidra.program.model.address import (  # pyright: ignore[reportMissingImports, reportMissingModuleSource, reportMissingTypeStubs]
        Address as GhidraAddress,
    )
    from ghidra.program.model.listing import (  # pyright: ignore[reportMissingImports, reportMissingModuleSource, reportMissingTypeStubs]
        Program as GhidraProgram,
    )
    from ghidra.program.model.mem import (  # pyright: ignore[reportMissingImports, reportMissingModuleSource, reportMissingTypeStubs]
        MemoryBlock as GhidraMemoryBlock,
    )
    from ghidra.program.model.symbol import (  # pyright: ignore[reportMissingImports, reportMissingModuleSource, reportMissingTypeStubs]
        Symbol as GhidraSymbol,
    )
    Symbol = GhidraSymbol

logger = logging.getLogger(__name__)


class MemoryUtil:
    """Utility functions for safe memory access operations."""

    @staticmethod
    def read_memory_bytes(program: GhidraProgram, address: GhidraAddress, length: int) -> bytes | None:
        """Read memory bytes safely with error handling.

        Args:
            program: The Ghidra program
            address: The address to read from
            length: Number of bytes to read

        Returns:
            The bytes read, or None if reading fails
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.read_memory_bytes")
        if program is None or address is None or length <= 0:
            return None

        try:
            memory = program.getMemory()
            if not memory.contains(address):
                return None

            # Ghidra getBytes expects a Java byte[]; create one via jpype, then convert result to Python bytes
            import jpype

            JByte = jpype.JClass("java.lang.Byte")
            buf = JByte[length]  # type: ignore

            n = memory.getBytes(address, buf)
            if n <= 0:
                return b""

            # Convert Java signed bytes to Python bytes
            return bytes(b & 0xFF for b in buf[:n])
        except Exception as e:
            logger.debug("Failed to read memory at %s: %s", address, e)
            return None

    @staticmethod
    def format_hex_string(data: bytes, bytes_per_line: int = 16) -> str:
        """Format bytes as a hex string with proper formatting.

        Args:
            data: The bytes to format
            bytes_per_line: Number of bytes per line

        Returns:
            Formatted hex string
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.format_hex_string")
        if not data:
            return ""

        lines = []
        for i in range(0, len(data), bytes_per_line):
            chunk = data[i : i + bytes_per_line]
            hex_part = " ".join(f"{b:02x}" for b in chunk)
            lines.append(hex_part)

        return "\n".join(lines)

    @staticmethod
    def byte_array_to_int_list(data: bytes) -> list[int]:
        """Convert bytes to a list of integers for JSON serialization.

        Args:
            data: The bytes to convert

        Returns:
            List of integers (0-255)
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.byte_array_to_int_list")
        return list(data)

    @staticmethod
    def find_block_by_name(program: GhidraProgram, block_name: str) -> GhidraMemoryBlock | None:
        """Find a memory block by name.

        Args:
            program: The Ghidra program
            block_name: Name of the memory block to find

        Returns:
            The memory block, or None if not found
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.find_block_by_name")
        if program is None or block_name is None:
            return None

        memory = program.getMemory()
        for block in memory.getBlocks():
            if block.getName() == block_name:
                return block

        return None

    @staticmethod
    def get_block_containing(program: GhidraProgram, address: GhidraAddress) -> GhidraMemoryBlock | None:
        """Get the memory block containing the given address.

        Args:
            program: The Ghidra program
            address: The address to check

        Returns:
            The memory block containing the address, or None
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.get_block_containing")
        if program is None or address is None:
            return None

        memory = program.getMemory()
        return memory.getBlock(address)

    @staticmethod
    def process_memory_in_chunks(
        program: GhidraProgram,
        start_address: GhidraAddress,
        total_length: int,
        chunk_size: int,
        processor: Callable[[bytes], None],
    ) -> None:
        """Process large memory regions in chunks to avoid memory issues.

        Args:
            program: The Ghidra program
            start_address: Starting address
            total_length: Total bytes to process
            chunk_size: Size of each chunk
            processor: Function to call for each chunk
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.process_memory_in_chunks")
        if program is None or start_address is None or total_length <= 0 or chunk_size <= 0:
            return

        current_address = start_address
        remaining = total_length

        while remaining > 0:
            # Calculate how much to read in this chunk
            to_read = min(chunk_size, remaining)

            # Read the chunk
            chunk = MemoryUtil.read_memory_bytes(program, current_address, to_read)
            if chunk is None:
                logger.warning("Failed to read memory chunk at %s", current_address)
                break

            if not chunk:
                logger.warning("Stopping chunked memory processing at %s due to empty read", current_address)
                break

            # Process the chunk
            try:
                processor(chunk)
            except Exception as e:
                logger.error("Error processing memory chunk at %s: %s", current_address, e)
                break

            # Move to next chunk
            try:
                current_address = current_address.add(len(chunk))
                remaining -= len(chunk)
            except Exception as e:
                logger.error("Error calculating next address: %s", e)
                break

    @staticmethod
    def is_address_in_executable_memory(program: GhidraProgram, address: GhidraAddress) -> bool:
        """Check if an address is in executable memory.

        Args:
            program: The Ghidra program
            address: The address to check

        Returns:
            True if the address is in executable memory
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.is_address_in_executable_memory")
        block = MemoryUtil.get_block_containing(program, address)
        return block is not None and block.isExecute()

    @staticmethod
    def is_address_in_writable_memory(program: GhidraProgram, address: GhidraAddress) -> bool:
        """Check if an address is in writable memory.

        Args:
            program: The Ghidra program
            address: The address to check

        Returns:
            True if the address is in writable memory
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.is_address_in_writable_memory")
        block = MemoryUtil.get_block_containing(program, address)
        return block is not None and block.isWrite()

    @staticmethod
    def get_memory_block_info(program: GhidraProgram, address: GhidraAddress) -> dict | None:
        """Get information about the memory block containing an address.

        Args:
            program: The Ghidra program
            address: The address to check

        Returns:
            Dictionary with block information, or None if not in any block
        """
        logger.debug("diag.enter %s", "mcp_utils/memory_util.py:MemoryUtil.get_memory_block_info")
        block = MemoryUtil.get_block_containing(program, address)
        if block is None:
            return None

        return {
            "name": block.getName(),
            "start": str(block.getStart()),
            "end": str(block.getEnd()),
            "size": block.getSize(),
            "readable": block.isRead(),
            "writable": block.isWrite(),
            "executable": block.isExecute(),
            "initialized": block.isInitialized(),
        }
